torch/csrc/autograd/variable_version.h

```diff
@@ -4,52 +4,95 @@
 
 namespace torch { namespace autograd {
 
+struct VersionBlock {
+  VersionBlock() : version(), live_refs(1) {}
+
+  // monotonically increasing version
+  std::atomic<int> version;
+  // number of references excluding SavedVariables
+  std::atomic<int> live_refs;
+};
+
+struct SavedVersion;
+
 struct VariableVersion {
-  VariableVersion() {
-    saved_ref = false;
-    version_block = new int[3];
-    version_block[0] = 0; // version
-    version_block[1] = 1; // refcount
-    version_block[2] = 1; // number of variables currently using the counter
-  };
+  VariableVersion() : version_block(std::make_shared<VersionBlock>()) {}
+  VariableVersion(const VariableVersion&) = delete;
+  VariableVersion(VariableVersion&&) = delete;
 
-  int operator++(int) { return version_block[0]++; }
+  ~VariableVersion() {
+    --version_block->live_refs;
+  }
+
+  // increment the version counter
+  void increment() { version_block->version++; }
 
-  int operator*() { return *version_block; }
+  // current version
+  int current_version() const { return version_block->version.load(); }
 
-  int var_refcnt() { return version_block[2]; }
+  // number of variables using this version counter (excludes SavedVariables)
+  int live_refs() const { return version_block->live_refs.load(); }
 
-  void join_with(VariableVersion &other) {
-    if (this == &other) {
-      return;
-    }
-    cleanup();
+  // creates a saved reference with the current version and the counter
+  inline SavedVersion save() const;
+
+  // Uses another variable's version counter. Used for variables which share storages
+  // NOTE: not thread-safe to call this from multiple threads without synchronization
```

Comment:
Just to make sure I understand, this is because the `shared_ptr` is not atomic, right?

Reply:
Yes, that's all